proto/ord_schema.proto

```diff
@@ -40,15 +37,16 @@ message Reaction {
 message ReactionIdentifier {
   // Possible identifier types are listed in an enum for extensibility
   enum IdentifierType {
-    REACTION_IDENTIFIER_TYPE_UNKNOWN = 0;  // Unspecified or custom.
-    REACTION_SMILES = 1;  // Reaction SMILES.
-    ATOM_MAPPED_SMILES = 2;  // Atom-mapped SMILES.
+    UNSPECIFIED = 0;
+    REACTION_SMILES = 1;
+    ATOM_MAPPED_SMILES = 2;
     RINCHI = 3;  // Reaction InChI.
     NAME = 4;  // Named reaction or reaction category.
   }
-  IdentifierType type = 1;
-  // Custom reaction identifier, used only when 'type' field is UNKNOWN.
-  string type_custom = 2;
+  oneof kind {
```

Comment:
Not being that familiar with ```oneof```s, will this make it slightly more verbose to check or existence? Or make a json export of a proto less self-explanatory? I'll defer to your judgement for these cases. I do agree that ```type``` and ```custom_type``` (or ```custom```) are mutually exclusive, so maybe it is worth enforcing that

Reply:
I also really like the enforcement of mutual exclusivity. Consider the following message:

```python
reaction = schema.Reaction()
reaction.identifiers.add(type='NAME', value='elephant toothpaste')
print(reaction)
>>> identifiers {
  type: NAME
  value: "elephant toothpaste"
}
```

* For JSON, the resulting dict will have only the `type` key, not `custom`.
* For SQL, both `type` and `custom` will show up as columns (and one will be NULL).
* For Python:
  ```python
  kind = reaction.identifiers[0].WhichOneof('kind')  # == 'type'`
  identifier_type = getattr(reaction.identifiers[0], kind)
  ```

Reply:
I like the JSON and SQL behavior, but that Python code hurts a little. In your example, would identifier_type be an integer if ```type``` is defined and a string if ```custom/type_custom``` is defined?

Reply:
A thought about forced exclusivity here: one of the uses of the string field would be to elaborate on the enum choice, since the enums aren't going to allow much precision. For example, the VesselMaterial "plastic" could mean any number of plastics, so we would want to be able to capture any additional (string) information. So maybe these fields shouldn't actually be exclusive? Current draft ```oneof``` fields where we might want to allow additional details are:
- preparation_type (we have separate "details" field at the moment)
- vessel_type
- vessel_material
- vessel_preparation
- temperature_control_type
- measurement_type
- pressure_control_type
- atmosphere_type (additional detail for OXYGEN might be the moisture level or grade, for example)
- etc.

Reply:
Interesting. I like the idea of being able to provide more details, but I also like that the oneof forces users to be explicit about their values. If we removed the exclusivity, I'd be worried that people might put in details about the glass they used without specifying GLASS as the material---we'd have no automatable way to check. Am I being too paranoid?

Would you support adding a `details` field to all of these like we do for preparation_type?

Reply:
I don't know if you're being too paranoid. I think that we could successfully encourage users to use the enums even if additional information is provided in the custom/details field.

I like the cleanliness of the ```details``` field for the ```Preparation``` message. I'm worried that adding a string field for details will become somewhat unwieldy, especially where we would need prefixes.

Reply:
What about something like this?

```proto
enum VesselMaterial {
  VESSEL_MATERIAL_UNSPECIFIED = 0;
  VESSEL_MATERIAL_CUSTOM = 1;
  GLASS = 2;
  POLYPROPYLENE = 3;
  PLASTIC = 4;
}
VesselMaterial material = 1;
string material_details = 2;
```

Then it's explicit that CUSTOM is being used and we can required material_details when that option is chosen.

(It might make sense to factor out a separate message to avoid the enum prefixes if we do this.)

Reply:
Would you propose we add a CUSTOM to all of the enums where we might want additional details?

Reply:
I think I'd do this for all the enums where a custom value is supported, and change `custom` to `details` everywhere. For enums that don't have a custom field but we might want details, I'd add a details field. So at the end every enum would have an accompanying details field, and some of them would have a CUSTOM option in the enum.

Reply:
That sounds reasonable to me. Also alleviates my ```oneof``` anxiety :)

Reply:
OK, great. I'll add another comment when I've updated.